lib/rubocop/cop/style/pattern_argument.rb

```diff
@@ -0,0 +1,83 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop looks for all? / any? / none? / one? calls where the block
+      # has a single statement and === has been invoked on the receiver so the
+      # === argument can be used as the predicate method argument.
+      #
+      # @example
+      #   # bad
+      #   %w[foo bar foobar].any? { |e| /oo/.match?(e) }
+      #
+      #   [1, 2, 3].any? { |e| (1..10).include?(e) }
+      #
+      #   [1, 2, 3].any? { |e| Set[1, 2, 3] === e }
+      #
+      #   [1, 2, 3].none? do |e|
+      #     Integer === e
+      #   end
+      #
+      #   # good
+      #   %w[foo bar foobar].any?(/oo/)
+      #
+      #   [1, 2, 3].any?(1..10)
+      #
+      #   [1, 2, 3].any?(Set[1, 2, 3])
+      #
+      #   [1, 2, 3].none?(Integer)
+      #
+      class PatternArgument < Base
+        include RangeHelp
+        extend AutoCorrector
+        extend TargetRubyVersion
+
+        minimum_target_ruby_version 2.5
+
+        def on_block(node)
+          pattern_argument_candidate?(node) do |method, pattern|
+            add_offense(
+              method.loc.selector,
+              message: format(MSG, method: method.method_name, pattern: pattern.source)
+            ) do |corrector|
+              corrector.replace(block_range_with_space(node), replacement(pattern))
+            end
+          end
+        end
+
+        private
+
+        MSG = 'Pass `%<pattern>s` as an argument to `%<method>s` instead of a block.'
+        RANGE_PATTERN = '(begin ({irange erange} {nil? _} {nil? _}))'
+
+        def_node_matcher :range?, <<~PATTERN
+          {(begin ({irange erange} _ _)) (send (const {nil? cbase} :Range) :new _ _)}
+        PATTERN
+        def_node_matcher :regexp?, <<~PATTERN
+          {(:regexp (:str _) _) (send (const {nil? cbase} :Regexp) :new _)}
+        PATTERN
+        def_node_matcher :set?, '(send (const {nil? cbase} :Set) {:[] :new} ...)'
+        def_node_matcher :pattern_argument_candidate?, <<~PATTERN
+          (block $({send csend} _ {:all? :any? :none? :one?}) (:args _) {
+            (send ${#range? #set?} {:=== :include? :member?} ...)
+            (send ${#regexp?} {:=== :=~ :match :match?} ...)
+            (send $(const nil? _) :=== _)
+            (send lvar :is_a? $(const nil? _))})
+        PATTERN
+
+        def replacement(pattern)
+          source = pattern.source
+          return source if RuboCop::AST::NodePattern.new(RANGE_PATTERN).match(pattern)
```

Comment:
You should never create `NodePattern`s on the fly, very ineficient. We should have an internal cop for this actually, I'll open an issue.

Reply:
What do you suggest? Checking the source by using a regular expression? Or using the `begin_type?` and `range_type?` predicates?

Reply:
I'm not sure I understand. You can use `def_node_matcher` (which does the compilation once instead of everytime like here), or `begin_type?`, or your previous string test...

Reply:
Ended up doing `pattern.begin_type? && [*pattern].first.range_type?` as I'm not sure what other things could just return true for `pattern.begin_type?`